### DISCUSSION TITLE
Using the dropdown for inputs & code modularity.

### DISCUSSION THREAD

**Discussion by nohehf** [janosh/svelte-multiselect#221] [Ideas]
Hi @janosh ! 
First, thank you for this amazing component. It is really a charm to work with it :)

I have had a case where I would like to use the dropdown for text input, not a select.

The existing dropdown multi-select and its features would make for a great, enhanced, and customizable version of a [datalist](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/datalist).

The current closest result to this you can get is achieved with something like:
```svelte
<MultiSelect
  options={suggestionOptions}
  allowUserOptions
  noMatchingOptionsMsg=""
  createOptionMsg=""
  maxSelect={1}
  required={1}
>
  <span let:option slot="option">
    Whatever custom slot
  </span>
</MultiSelect>
```
But this is not really usable as it has some perks, as it doesn't feel like an input:
- Text is not editable without removing the whole element
- "Card" look
- ...
<img width="654" alt="image" src="https://user-images.githubusercontent.com/64989428/229223242-193409c8-a3ed-4534-99af-1ba884898072.png">

It would be nice to separate the Multiselect component into multiple components. It would allow the drop-down to be used for inputs and probably serve other use cases.

I know this is not the standard use case for the component, though it would make the library even more versatile and improve the overall quality/clarity.

If this seems useful to you, I can start working on a MR (tho I cannot promise any ETA). The first step would be to split the component, then use adapt the dropdown to make it usable outside of the multiselect, and last maybe export an Input component.

Let me know what you think and thanks again for the amazing work!

PS: I have a lot of TS errors in my editor related to `Options` type, in the form of:
```
Argument of type '...' is not assignable to parameter of type 'Option'.
  '...' is assignable to the constraint of type 'Option', but 'Option' could be instantiated with a different subtype of constraint 'Option'.
 ```

PS2: Just wondering, why isn't the lock file committed ?

**Comment by janosh** [comment]
Thanks for bringing this up! And for #220! 👍 

> It would be nice to separate the Multiselect component into multiple components. It would allow the drop-down to be used for inputs and probably serve other use cases.

I've thought about this once or twice but so far have lacked motivation because I didn't need it myself. But if you want to tackle this, feel free. PRs welcome!

> PS: I have a lot of TS errors in my editor related to Options type, in the form of:

Do you mean using when using the component or developing it? Either way, I'd appreciate any and all fixes!

> PS2: Just wondering, why isn't the lock file committed ?

Life on the edge. :)
I never encountered a need for them yet with Svelte. No clashing transitive dep version. Maybe because the deps tree of my Svelte projects is more shallow than with other frameworks. So all the lock file did for me was clutter up git history.

**Comment by uh-zuh** [comment]
Any update?
I need such input with dropdown too.

**Reply by janosh** [reply]
i think https://github.com/janosh/svelte-multiselect/pull/416 adds what you're looking for. could you give https://multiselect.janosh.dev/input-dropdown a try and let me know if anything's missing?

**Reply by uh-zuh** [reply]
Good!
Additionally I need to be able to see all options when I click on left-side-button with carets to be able to select another option and additionally I want to see which option is selected (if selected) in this dropdown list.

**Reply by janosh** [reply]
see https://github.com/janosh/svelte-multiselect/commit/68e5bef88c50123ab6118a25214c64cd3650cca6

**Reply by uh-zuh** [reply]
Dropdown opens only if item is selected but when custom value is typed then only "No matching options" shows but I need ability to select another option in this case too.

**Reply by janosh** [reply]
398b819 caret-click now only shows the full option list when the current typed input has no matches. partial matches still stay filtered, and committed selections still reopen with the selected option marked

**Reply by uh-zuh** [reply]
I need ability to see all options in any case:
- Some item is selected
- Few characters have typed and there are some matching items in dropdown list
- A lot of characters have typed and there are no matching items in dropdown list

I other words I need to see full list when I click on carets regardless of state of component.

Additionally I prefer to hide dropdown list on second click on carets.

**Reply by janosh** [reply]
0ca9df8 in `input-dropdown` mode, the caret now works as a browse-all toggle. If the dropdown is closed, clicking the caret opens the full option list regardless of whether an option is selected, the typed text has matches, or the typed text has no matches. if the dropdown is already open, clicking the caret closes it.
i kept typing behavior unchanged: typing in the input still filters suggestions normally.

**Reply by uh-zuh** [reply]
Perfect! Thank you so much!

**Reply by uh-zuh** [reply]
When you are planning release?

**Reply by janosh** [reply]
[v11.7.1](https://github.com/janosh/svelte-multiselect/releases/tag/v11.7.1)

**Reply by uh-zuh** [reply]
Thank you!
